### ISSUE TITLE
server: disk IO stat gathering crashes in tests on macOS

### ISSUE THREAD

**Issue by vilterp** [cockroachdb/cockroach#27867]
**Repro**

1. Run `make testshort PKG='./pkg/sql/opt/exec/execbuilder' TESTFLAGS='-count=1'`

**Expected**

Tests pass.

**Actual**

```
SIGILL: illegal instruction
PC=0x7fff54fe935d m=5 sigcode=1
```
...
```
goroutine 83383 [syscall]:
github.com/cockroachdb/cockroach/vendor/github.com/shirou/gopsutil/disk._Cfunc_ReadDiskInfo(0xc4243b00a0, 0xc400000000)
	_cgo_gotypes.go:99 +0x49
github.com/cockroachdb/cockroach/vendor/github.com/shirou/gopsutil/disk.IOCountersWithContext(0x628a4c0, 0xc42173ffb0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/vendor/github.com/shirou/gopsutil/disk/disk_darwin_cgo.go:59 +0x153
github.com/cockroachdb/cockroach/pkg/server/status.getSummedDiskCounters(0x628a4c0, 0xc42173ffb0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/server/status/runtime.go:463 +0x88
github.com/cockroachdb/cockroach/pkg/server/status.MakeRuntimeStatSampler(0x628a4c0, 0xc42173ffb0, 0xc425fd6050, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/server/status/runtime.go:304 +0x529
github.com/cockroachdb/cockroach/pkg/server.NewServer(0xc4244aa690, 0xc423890000, 0xbebc200, 0xf, 0x4008000000000000, 0xc424f24100, 0x0, 0x0, 0xc42548e0e0, 0x1, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:434 +0x1e4c
github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start(0xc42548e150, 0x6266560, 0xc4214d0180, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, 0x0, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:336 +0xc9
github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartServerRaw(0x6266560, 0xc4214d0180, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:206 +0x109
github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartServer(0x62ae680, 0xc4252af860, 0x6266560, 0xc4214d0180, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:174 +0x63
github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).doAddServer(0xc42548e000, 0x62ae680, 0xc4252af860, 0x6266560, 0xc4214d0180, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:251 +0x123
github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster(0x62ae680, 0xc4252af860, 0x1, 0x6266560, 0xc4214d00c0, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:157 +0x5d9
github.com/cockroachdb/cockroach/pkg/testutils/testcluster.testClusterFactoryImpl.StartTestCluster(0x62ae680, 0xc4252af860, 0x1, 0x6266560, 0xc4214d00c0, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:588 +0x8a
github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartTestCluster(0x62ae680, 0xc4252af860, 0x1, 0x6266560, 0xc4214d00c0, 0x0, 0x0, 0x62644a0, 0xc423a46000, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_cluster_shim.go:120 +0xa7
github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).setup(0xc420884000, 0x5dd915b, 0x9, 0x1, 0x0, 0x5dcf6ff, 0x2, 0x5dcfd4a, 0x3, 0x0, ...)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:981 +0x472
github.com/cockroachdb/cockroach/pkg/sql/logictest.RunLogicTest.func1.1(0xc4252af860)
	/Users/vilterp/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:2075 +0x16f
testing.tRunner(0xc4252af860, 0xc425b1a8a0)
	/usr/local/Cellar/go/1.10/libexec/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
	/usr/local/Cellar/go/1.10/libexec/src/testing/testing.go:824 +0x2e0
```

Only the tests in this package fail. This failure doesn't happen on Linux.

**Diagnosis**

tldr: library we use to gather disk IO stats for new Hardware dashboard in admin UI (introduced in #27137) is poorly written (uses shared mutable globals).

The execbuilder tests run with very high concurrency (higher than any other tests, we think), and create a lot of `server.Server` structs. Each one starts a `RuntimeStatSampler`, which every 10 seconds calls into `ReadDiskInfo` in http://github.com/shirou/gopsutil, which uses [these global variables](https://github.com/shirou/gopsutil/blob/master/disk/disk_darwin.h#L34-L39). Under high enough concurrency, two threads use the global variables at the same time, causing some kind of corruption and crash.

The library's Linux implementation is different (no CGo code) so it doesn't fail in this way. A normal Cockroach process also doesn't fail in this way, since there's only one thread calling into gopsutil.

**Remedy**

Prometheus's NodeExporter tool uses http://github.com/lufia/iostat, which appears to make the same CGo calls into `IOKit`, but without using global variables. Hopefully it will be fine under contention.

Thanks @benesch and @andreimatei for debugging help.

**Comment by vilterp** [comment]
Complication: `iostat` doesn't implement disk stats for linux https://github.com/lufia/iostat/blob/master/iostat_linux.go#L10-L12

**Comment by vilterp** [comment]
Filed https://github.com/shirou/gopsutil/issues/560

Using node_exporter would be nice, but its API looks very tied to Prometheus's collection system, not for use as a library — you can't just get the values of the metrics you want; it pushes updates to you over a channel https://github.com/prometheus/node_exporter/blob/24a28fcc9e9123b3b39058918f98b64a1938a0a9/collector/collector.go#L143-L146

**Comment by petermattis** [comment]
> Using node_exporter would be nice, but its API looks very tied to Prometheus's collection system, not for use as a library

`node_exporter` is open-source. We can fork it and adjust the API to whatever we need. I'm not saying we should do this, just that it is an option to consider.
